Core/Configuration/JsonConfiguration.cs

```diff
@@ -0,0 +1,419 @@
+                    {
+                        Host = host,
+                        Token = token
+                    });
+                }
+
+                SaveConfig();
+            }
+        }
+
+        // <summary>
+        // Save the JSON configuration file. Only call this while you own _lock.
+        // </summary>
+        private static void SaveConfig()
+        {
+            string json = JsonConvert.SerializeObject(config, Formatting.Indented);
+            File.WriteAllText(configFile, json);
+        }
+
+        // <summary>
+        // Load the JSON configuration file. This will replace the current state.
+        // Only call this while you own _lock.
+        //
+        // If the configuration file does not exist, this will create it and then
+        // try to populate it with values in the registry left from the old system.
+        // </summary>
+        private void LoadConfig()
+        {
+            try
+            {
+                string json = File.ReadAllText(configFile);
+                config = JsonConvert.DeserializeObject<Config>(json);
+
+                if (config == null)
+                {
+                    config = new Config();
+                }
+
+                if (config.KspInstances == null)
+                {
+                    config.KspInstances = new List<KspInstance>();
+
+                }
+
+                if (config.AuthTokens == null)
+                {
+                    config.AuthTokens = new List<AuthToken>();
+                }
+            }
+            catch (Exception ex) when (ex is FileNotFoundException || ex is DirectoryNotFoundException)
+            {
+                // This runs if the configuration does not exist. We will create a new configuration and
+                // try to migrate from the registry.
+                config = new Config();
+                config.KspInstances = new List<KspInstance>();
+                config.AuthTokens = new List<AuthToken>();
+
+                // Ensure the directory exists
+                new FileInfo(configFile).Directory.Create();
+
+                // Write the configuration to the disk
+                SaveConfig();
+
+#if !NETSTANDARD
+                // If we are not running on .NET Standard, try to migrate from the real registry
+                Migrate();
+#endif
+            }
+        }
+
+        // <summary>
+        // Copy the configuration from the registry here.
+        // </summary>
+        private void Migrate()
+        {
+            RegistryConfiguration registry = new RegistryConfiguration();
+
+            var instances = registry.GetInstances();
+            lock (_lock)
+            {
+                config.KspInstances = instances.Select(instance => new KspInstance
+                {
+                    Name = instance.Item1,
+                    Path = instance.Item2
+                }).ToList();
+
+                SaveConfig();
```

Comment:
Shouldn't SaveConfig be at the end of this function, after all the properties are set?

Reply:
Currently SaveConfig() gets called after every write operation, which is analogous to how the registry works, where all changes are immediately persistent. I have the SaveConfig() here to go along with that pattern. I could probably change Migrate() to only save once at the end, although I'm not sure if the added complexity is justified.

Reply:
I have modified Migrate so that it only calls SaveConfig once at the end of the function.